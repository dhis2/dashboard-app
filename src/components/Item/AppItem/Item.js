import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import NotInterestedIcon from '@material-ui/icons/NotInterested';

import { FILTER_USER_ORG_UNIT } from '../../../actions/itemFilter';
import { sGetItemFilterRoot } from '../../../reducers/itemFilter';
import ItemHeader from '../ItemHeader';
import Line from '../../../widgets/Line';

const getIframeSrc = (appDetails, item, itemFilter) => {
    let iframeSrc = `${appDetails.launchUrl}?dashboardItemId=${item.id}`;

    if (
        itemFilter &&
        itemFilter[FILTER_USER_ORG_UNIT] &&
        itemFilter[FILTER_USER_ORG_UNIT].length
    ) {
        const ouIds = itemFilter[FILTER_USER_ORG_UNIT].map(
            ouPath => ouPath.split('/').slice(-1)[0]
        );

        iframeSrc += `&userOrgUnit=${ouIds.join(',')}`;
    }

    return iframeSrc;
};

const AppItem = ({ item, itemFilter }, context) => {
    let appDetails;

    const appKey = item.appKey;

    if (appKey) {
        appDetails = context.d2.system.installedApps.find(
            app => app.key === appKey
        );
    }

    return appDetails && appDetails.name && appDetails.launchUrl ? (
        <Fragment>
            <ItemHeader title={appDetails.name} />
            <Line />
            <iframe
                title={appDetails.name}
                src={getIframeSrc(appDetails, item, itemFilter)}
                className="dashboard-item-content"
                style={{ border: 'none' }}
            />
        </Fragment>
    ) : (
        <Fragment>
            <ItemHeader title={`${appKey} app not found`} />
            <Line />
            <div
                className="dashboard-item-content"
                style={{
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    height: '90%',
                }}
            >
                <NotInterestedIcon
                    color="disabled"
                    disabled
                    style={{ width: 100, height: 100, align: 'center' }}
                />
            </div>
        </Fragment>
    );
};

AppItem.contextTypes = {
    d2: PropTypes.object,
};

const mapStateToProps = state => ({
    itemFilter: sGetItemFilterRoot(state),
});

export default connect(mapStateToProps)(AppItem);
