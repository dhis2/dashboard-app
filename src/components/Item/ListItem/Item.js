import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { colors } from '@dhis2/ui-core'
import DescriptionIcon from '../../../icons/Description'

import DeleteIcon from '../../../icons/Delete'
import Line from '../../../widgets/Line'
import { itemTypeMap, getItemUrl } from '../../../modules/itemTypes'
import { orArray } from '../../../modules/util'
import { tRemoveListItemContent } from './actions'
import ItemHeader from '../ItemHeader'
import classes from './Item.module.css'

const getItemTitle = item => itemTypeMap[item.type].pluralTitle

const getContentItems = item =>
    orArray(item[itemTypeMap[item.type].propName]).filter(
        (item, index, array) =>
            array.findIndex(el => el.id === item.id) === index
    )

const ListItem = (props, context) => {
    const { item, editMode, tRemoveListItemContent } = props
    const contentItems = getContentItems(item)

    const getLink = contentItem => {
        const deleteButton = (
            <button
                className={classes.deletebutton}
                onClick={() => tRemoveListItemContent(item, contentItem)}
            >
                <DeleteIcon className={classes.deleteicon} />
            </button>
        )

        return (
            <>
                <a
                    className={classes.link}
                    style={{ color: colors.grey900 }}
                    href={getItemUrl(item.type, contentItem, context.d2)}
                >
                    {contentItem.name}
                </a>
                {editMode ? deleteButton : null}
            </>
        )
    }

    return (
        <>
            <ItemHeader title={getItemTitle(item)} itemId={item.id} />
            <Line />
            <div className="dashboard-item-content">
                <ul className={classes.list}>
                    {contentItems.map(contentItem => (
                        <li className={classes.item} key={contentItem.id}>
                            <DescriptionIcon className={classes.itemicon} />
                            {getLink(contentItem)}
                        </li>
                    ))}
                </ul>
            </div>
        </>
    )
}

ListItem.propTypes = {
    editMode: PropTypes.bool,
    item: PropTypes.object,
    tRemoveListItemContent: PropTypes.func,
}

ListItem.contextTypes = {
    d2: PropTypes.object,
}

export default connect(null, {
    tRemoveListItemContent,
})(ListItem)
