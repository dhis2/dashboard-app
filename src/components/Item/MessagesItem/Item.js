import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import { useConfig } from '@dhis2/app-runtime'
import ItemHeader from '../ItemHeader/ItemHeader'
import Line from '../../../widgets/Line'
import { useUserSettings } from '../../UserSettingsProvider'

import { sGetMessagesRoot } from '../../../reducers/messages'
import { formatDate } from '../../../modules/util'
import { isViewMode } from '../../Dashboard/dashboardModes'

import classes from './styles/Item.module.css'
import './MessagesItem.css'

const PRIVATE = 'PRIVATE'

const messageTypes = {
    [PRIVATE]: 'Private',
    VALIDATION_RESULT: 'Validation',
    TICKET: 'Ticket',
    SYSTEM: 'System',
}

const MessagesItem = ({ messages, item, dashboardMode }) => {
    const { baseUrl } = useConfig()
    const { userSettings } = useUserSettings()

    const getMessageHref = msg => {
        const msgIdentifier = msg ? `#/${msg.messageType}/${msg.id}` : ''
        return `${baseUrl}/dhis-web-messaging/${msgIdentifier}`
    }

    const getMessageSender = msg => {
        const latestMsg = msg.messages.slice(-1)[0]
        return latestMsg.sender ? latestMsg.sender.displayName : ''
    }

    const getMessageItems = () => {
        const modeClass = isViewMode(dashboardMode) ? 'view' : null

        return messages.map(msg => {
            const redirectToMsg = () => {
                if (isViewMode(dashboardMode)) {
                    document.location.href = getMessageHref(msg)
                }
            }

            const sender =
                msg.messageType === PRIVATE
                    ? getMessageSender(msg)
                    : messageTypes[msg.messageType]

            const readClass = !msg.read ? 'unread' : null
            const latestMsg = msg.messages.slice(-1)[0]
            const msgDate = latestMsg.lastUpdated

            return (
                <li
                    className={`message-item ${modeClass}`}
                    key={msg.id}
                    onClick={redirectToMsg}
                >
                    <p className={`message-title ${readClass}`}>
                        {msg.displayName} ({msg.messageCount})
                    </p>
                    <p className={classes.sender}>
                        {sender} -{' '}
                        {formatDate(msgDate, userSettings.keyUiLocale)}
                    </p>
                    <p className={classes.snippet}>{latestMsg.text}</p>
                </li>
            )
        })
    }

    return (
        <>
            <ItemHeader
                title={i18n.t('Messages')}
                itemId={item.id}
                dashboardMode={dashboardMode}
                isShortened={item.shortened}
            />
            <Line />
            {messages.length > 0 && (
                <div className="dashboard-item-content">
                    <ul className={classes.list}>{getMessageItems()}</ul>
                    <div className={classes.seeAll}>
                        <a href={getMessageHref()}>
                            {i18n.t('See all messages')}
                        </a>
                    </div>
                </div>
            )}
        </>
    )
}

MessagesItem.propTypes = {
    dashboardMode: PropTypes.string,
    item: PropTypes.object,
    messages: PropTypes.array,
}

const mapStateToProps = state => {
    return {
        messages: Object.values(sGetMessagesRoot(state)),
    }
}

const MessagesContainer = connect(mapStateToProps, null)(MessagesItem)

export default MessagesContainer
