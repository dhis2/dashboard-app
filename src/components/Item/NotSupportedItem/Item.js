import React, { Fragment } from 'react';
import ItemHeader from '../ItemHeader';
import SvgIcon from 'd2-ui/lib/svg-icon/SvgIcon';

const NotSupportedItem = props => (
    <Fragment>
        <ItemHeader title={`Item type not supported: ${props.item.type}`} />
        <div
            style={{
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                height: '90%',
            }}
        >
            <SvgIcon
                icon="NotInterested"
                style={{ width: 100, height: 100, align: 'center' }}
                disabled
            />
        </div>
    </Fragment>
);

export default NotSupportedItem;
