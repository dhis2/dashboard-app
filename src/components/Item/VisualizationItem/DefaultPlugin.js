import React, { Component } from 'react';
import PropTypes from 'prop-types';

import i18n from '@dhis2/d2-i18n';

import * as pluginManager from './plugin';
import { getBaseUrl, orObject } from '../../../modules/util';
import { getGridItemDomId } from '../../ItemGrid/gridUtil';

const pluginCredentials = d2 => {
    return {
        baseUrl: getBaseUrl(d2),
        auth: d2.Api.getApi().defaultHeaders.Authorization,
    };
};

class DefaultPlugin extends Component {
    pluginCredentials = null;

    constructor(props, context) {
        super(props);

        this.d2 = context.d2;
    }

    shouldPluginReload = prevProps => {
        // TODO - fix this hack, to handle bug with multiple
        // rerendering while switching between dashboards.
        //
        // To determine if the rendering is happening because of a
        // dashboard switch, check if the item reference has changed.
        const reloadAllowed = this.props.item === prevProps.item;

        const filtersChanged = prevProps.itemFilters !== this.props.itemFilters;
        const vis = orObject(this.props.visualization);
        const prevVis = orObject(prevProps.visualization);
        const visChanged =
            vis.id !== prevVis.id || vis.activeType !== prevVis.activeType;

        return reloadAllowed && (visChanged || filtersChanged);
    };

    reloadPlugin = prevProps => {
        if (
            pluginManager.pluginIsAvailable(
                this.props.item,
                this.props.visualization
            ) &&
            this.shouldPluginReload(prevProps)
        ) {
            const prevVis = orObject(prevProps.visualization);
            const currentVis = this.props.visualization;

            const useActiveType =
                currentVis.activeType !== prevVis.activeType ||
                currentVis.activeType !== this.props.item.type;

            if (
                useActiveType ||
                this.props.itemFilters !== prevProps.itemFilters
            ) {
                pluginManager.unmount(
                    this.props.item,
                    prevVis.activeType || this.props.item.type
                );

                pluginManager.load(
                    this.props.item,
                    this.props.visualization,
                    this.pluginCredentials,
                    useActiveType ? currentVis.activeType : null
                );
            }
        }
    };

    componentDidMount() {
        this.pluginCredentials = pluginCredentials(this.d2);

        if (
            pluginManager.pluginIsAvailable(
                this.props.item,
                this.props.visualization
            )
        ) {
            pluginManager.load(
                this.props.item,
                this.props.visualization,
                this.pluginCredentials,
                !this.props.editMode ? this.getActiveType() : null
            );
        }
    }

    componentDidUpdate(prevProps) {
        this.reloadPlugin(prevProps);
    }

    getActiveType = () =>
        this.props.visualization.activeType || this.props.item.type;

    render() {
        const { classes, item, visualization, style } = this.props;
        const pluginIsAvailable = pluginManager.pluginIsAvailable(
            item,
            visualization
        );

        return pluginIsAvailable ? (
            <div id={getGridItemDomId(item.id)} style={style} />
        ) : (
            <div className={classes.textDiv}>
                {i18n.t('Unable to load the plugin for this item')}
            </div>
        );
    }
}

DefaultPlugin.contextTypes = {
    d2: PropTypes.object,
};

DefaultPlugin.propTypes = {
    style: PropTypes.object,
    item: PropTypes.object,
    itemFilters: PropTypes.object,
    visualization: PropTypes.object,
};

DefaultPlugin.defaultProps = {
    style: {},
    item: {},
    itemFilters: {},
    visualization: {},
};

export default DefaultPlugin;
