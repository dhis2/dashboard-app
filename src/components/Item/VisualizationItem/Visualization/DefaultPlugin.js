import { useConfig } from '@dhis2/app-runtime'
import { useD2 } from '@dhis2/app-runtime-adapter-d2'
import PropTypes from 'prop-types'
import React, { useEffect, useRef } from 'react'
import getVisualizationContainerDomId from '../getVisualizationContainerDomId.js'
import { load, unmount } from './plugin.js'

const DefaultPlugin = ({
    item,
    activeType,
    filterVersion,
    mapViewCount,
    visualization,
    options,
    style,
}) => {
    const { d2 } = useD2()
    const { baseUrl } = useConfig()
    const credentials = {
        baseUrl,
        auth: d2.Api.getApi().defaultHeaders.Authorization,
    }

    const prevItem = useRef()
    const prevActiveType = useRef()
    const prevFilterVersion = useRef()
    const prevMapViewCount = useRef()
    const prevVisualization = useRef()

    useEffect(() => {
        load(item, visualization, {
            credentials,
            activeType,
            d2,
            options,
        })

        prevItem.current = item
        prevActiveType.current = activeType
        prevFilterVersion.current = filterVersion
        prevMapViewCount.current = mapViewCount
        prevVisualization.current = visualization

        return () => unmount(item, item.type || activeType)
    }, [])

    useEffect(() => {
        if (
            prevItem.current === item &&
            (prevActiveType.current !== activeType ||
                prevFilterVersion.current !== filterVersion ||
                prevMapViewCount.current < mapViewCount ||
                prevVisualization.current !== visualization)
        ) {
            /* Item is the same but type or filters has changed
             * or map was previously loaded with fewer mapViews
             * so necessary to reload
             */

            load(item, visualization, {
                credentials,
                activeType,
                d2,
                options,
            })
        }

        prevItem.current = item
        prevActiveType.current = activeType
        prevFilterVersion.current = filterVersion
        prevVisualization.current = visualization
    }, [item, visualization, activeType, filterVersion, mapViewCount])

    return <div id={getVisualizationContainerDomId(item.id)} style={style} />
}

DefaultPlugin.propTypes = {
    activeType: PropTypes.string,
    filterVersion: PropTypes.string,
    item: PropTypes.object,
    mapViewCount: PropTypes.number,
    options: PropTypes.object,
    style: PropTypes.object,
    visualization: PropTypes.object,
}

DefaultPlugin.defaultProps = {
    style: {},
    item: {},
    options: {},
    visualization: {},
}

export default DefaultPlugin
