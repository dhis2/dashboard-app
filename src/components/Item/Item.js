import React from 'react';

import VisualizationItem from './VisualizationItem/Item';
import MessagesItem from './MessagesItem/Item';
import NotSupportedItem from './NotSupportedItem/Item';
import ListItem from './ListItem/Item';
import TextItem from './TextItem/Item';
import AppItem from './AppItem/Item';
import SpacerItem from './SpacerItem/Item';
import {
    APP,
    REPORT_TABLE,
    CHART,
    MAP,
    EVENT_CHART,
    EVENT_REPORT,
    MESSAGES,
    REPORTS,
    RESOURCES,
    USERS,
    TEXT,
    SPACER,
} from '../../modules/itemTypes';
import { DEFAULT_STATE_ITEM_FILTERS } from '../../reducers/itemFilters';

const getGridItem = type => {
    switch (type) {
        case REPORT_TABLE:
        case CHART:
        case MAP:
        case EVENT_CHART:
        case EVENT_REPORT:
            return VisualizationItem;
        case MESSAGES:
            return MessagesItem;
        case REPORTS:
        case RESOURCES:
        case USERS:
            return ListItem;
        case TEXT:
            return TextItem;
        case SPACER:
            return SpacerItem;
        case APP:
            return AppItem;
        default:
            return NotSupportedItem;
    }
};

export const Item = props => {
    const GridItem = getGridItem(props.item.type);

    return (
        <GridItem
            item={props.item}
            editMode={props.editMode}
            itemFilters={
                props.editMode ? DEFAULT_STATE_ITEM_FILTERS : props.itemFilters
            }
            onToggleItemExpanded={props.onToggleItemExpanded}
        />
    );
};
