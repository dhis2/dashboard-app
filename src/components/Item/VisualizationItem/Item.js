import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import uniqueId from 'lodash/uniqueId'
import VisualizationPlugin from '@dhis2/data-visualizer-plugin'
import i18n from '@dhis2/d2-i18n'

import DefaultPlugin from './DefaultPlugin'
import FatalErrorBoundary from './FatalErrorBoundary'
import ItemHeader, { HEADER_MARGIN_HEIGHT } from '../ItemHeader'
import ItemHeaderButtons from './ItemHeaderButtons'
import ItemFooter from './ItemFooter'
import * as pluginManager from './plugin'
import { sGetVisualization } from '../../../reducers/visualizations'
import { sGetItemFiltersRoot } from '../../../reducers/itemFilters'
import {
    acAddVisualization,
    acSetActiveVisualizationType,
} from '../../../actions/visualizations'
import {
    VISUALIZATION,
    MAP,
    CHART,
    REPORT_TABLE,
} from '../../../modules/itemTypes'
import memoizeOne from '../../../modules/memoizeOne'

import { getVisualizationConfig } from './plugin'
import LoadingMask from './LoadingMask'
import { ITEM_CONTENT_PADDING_BOTTOM } from '../../ItemGrid/ItemGrid'

import classes from './styles/Item.module.css'

export class Item extends Component {
    state = {
        showFooter: false,
        configLoaded: false,
        pluginIsLoaded: false,
    }

    constructor(props, context) {
        super(props)

        this.d2 = context.d2

        this.contentRef = React.createRef()
        this.headerRef = React.createRef()

        this.memoizedApplyFilters = memoizeOne(this.applyFilters)

        this.memoizedGetVisualizationConfig = memoizeOne(getVisualizationConfig)

        this.memoizedGetContentStyle = memoizeOne(this.getContentStyle)
    }

    async componentDidMount() {
        this.props.onVisualizationLoaded(
            // TODO do not call fetch on the pluginManager, do it here as the manager will eventually be removed...
            await pluginManager.fetch(this.props.item)
        )

        this.setState({
            configLoaded: true,
        })
    }

    componentDidUpdate(prevProps, prevState) {
        if (
            prevState.pluginIsLoaded &&
            (prevProps.visualization !== this.props.visualization ||
                prevProps.itemFilters !== this.props.itemFilters)
        ) {
            this.setState({
                pluginIsLoaded: false,
            })
        }
    }

    applyFilters = (visualization, filters) => {
        if (!Object.keys(filters).length) {
            return visualization
        }

        // deep clone objects in filters to avoid changing the visualization in the Redux store
        const visRows = visualization.rows.map(obj => ({ ...obj }))
        const visColumns = visualization.columns.map(obj => ({ ...obj }))
        const visFilters = visualization.filters.map(obj => ({ ...obj }))

        Object.keys(filters).forEach(dimensionId => {
            if (filters[dimensionId]) {
                let dimensionFound = false

                ;[visRows, visColumns, visFilters].forEach(dimensionObjects =>
                    dimensionObjects
                        .filter(obj => obj.dimension === dimensionId)
                        .forEach(obj => {
                            dimensionFound = true
                            obj.items = filters[dimensionId]
                        })
                )

                // add dimension to filters if not already present elsewhere
                if (!dimensionFound) {
                    visFilters.push({
                        dimension: dimensionId,
                        items: filters[dimensionId],
                    })
                }
            }
        })

        return {
            ...visualization,
            rows: visRows,
            columns: visColumns,
            filters: visFilters,
        }
    }

    getUniqueKey = memoizeOne(() => uniqueId())

    pluginCredentials = null

    getPluginComponent = () => {
        const activeType = this.getActiveType()
        const visualization = this.memoizedGetVisualizationConfig(
            this.props.visualization,
            this.props.item.type,
            activeType
        )

        if (!visualization) {
            return (
                <div className={classes.textDiv}>
                    {i18n.t('No data to display')}
                </div>
            )
        }

        const calculatedHeight =
            this.props.item.originalHeight -
            this.headerRef.current.clientHeight -
            HEADER_MARGIN_HEIGHT -
            ITEM_CONTENT_PADDING_BOTTOM

        const props = {
            ...this.props,
            visualization,
            classes,
            style: this.memoizedGetContentStyle(
                calculatedHeight,
                this.contentRef ? this.contentRef.offsetHeight : null,
                this.props.editMode
            ),
        }

        switch (activeType) {
            case VISUALIZATION:
            case CHART:
            case REPORT_TABLE: {
                return (
                    <>
                        {!this.state.pluginIsLoaded && (
                            <div style={props.style}>
                                <LoadingMask />
                            </div>
                        )}
                        <VisualizationPlugin
                            d2={this.d2}
                            visualization={this.memoizedApplyFilters(
                                visualization,
                                props.itemFilters
                            )}
                            onLoadingComplete={this.onLoadingComplete}
                            forDashboard={true}
                            style={props.style}
                        />
                    </>
                )
            }
            case MAP: {
                if (props.item.type === MAP) {
                    // apply filters only to thematic and event layers
                    // for maps AO
                    const mapViews = props.visualization.mapViews.map(obj => {
                        if (
                            obj.layer.includes('thematic') ||
                            obj.layer.includes('event')
                        ) {
                            return this.memoizedApplyFilters(
                                obj,
                                props.itemFilters
                            )
                        }

                        return obj
                    })

                    props.visualization = {
                        ...props.visualization,
                        mapViews,
                    }
                } else {
                    // this is the case of a non map AO passed to the maps plugin
                    // due to a visualization type switch in dashboard item
                    // maps plugin takes care of converting the AO to a suitable format
                    props.visualization = this.memoizedApplyFilters(
                        props.visualization,
                        props.itemFilters
                    )
                }

                props.options = {
                    ...props.options,
                    hideTitle: true,
                }

                return <DefaultPlugin {...props} />
            }
            default: {
                props.visualization = this.memoizedApplyFilters(
                    props.visualization,
                    props.itemFilters
                )

                return <DefaultPlugin {...props} />
            }
        }
    }

    onLoadingComplete = () => {
        this.setState({
            pluginIsLoaded: true,
        })
    }

    onToggleFooter = () => {
        this.setState(
            { showFooter: !this.state.showFooter },
            this.props.onToggleItemExpanded(this.props.item.id)
        )
    }

    onSelectActiveType = type => {
        if (type === this.getActiveType()) {
            return
        }

        pluginManager.unmount(this.props.item, this.getActiveType())

        this.props.onSelectActiveType(this.props.visualization.id, type)
    }

    getActiveType = () =>
        this.props.visualization.activeType || this.props.item.type

    pluginIsAvailable = () =>
        pluginManager.pluginIsAvailable(
            this.props.item,
            this.props.visualization
        )

    getContentStyle = (calculatedHeight, measuredHeight, editMode) => {
        const height = editMode
            ? measuredHeight || calculatedHeight
            : calculatedHeight

        return { height }
    }

    render() {
        const { item, editMode, itemFilters } = this.props
        const { showFooter } = this.state

        const actionButtons = (
            <ItemHeaderButtons
                item={item}
                visualization={this.props.visualization}
                onSelectActiveType={this.onSelectActiveType}
                onToggleFooter={this.onToggleFooter}
                d2={this.d2}
                activeType={this.getActiveType()}
                activeFooter={this.state.showFooter}
            />
        )

        return (
            <>
                <ItemHeader
                    title={pluginManager.getName(item)}
                    itemId={item.id}
                    actionButtons={actionButtons}
                    ref={this.headerRef}
                />
                <FatalErrorBoundary>
                    <div
                        key={this.getUniqueKey(itemFilters)}
                        className="dashboard-item-content"
                        ref={ref => (this.contentRef = ref)}
                    >
                        {this.state.configLoaded && this.getPluginComponent()}
                    </div>
                </FatalErrorBoundary>
                {!editMode && showFooter ? <ItemFooter item={item} /> : null}
            </>
        )
    }
}

Item.contextTypes = {
    d2: PropTypes.object,
}

Item.propTypes = {
    editMode: PropTypes.bool,
    item: PropTypes.object,
    itemFilters: PropTypes.object,
    visualization: PropTypes.object,
    onSelectActiveType: PropTypes.func,
    onToggleItemExpanded: PropTypes.func,
    onVisualizationLoaded: PropTypes.func,
}

Item.defaultProps = {
    item: {},
    editMode: false,
    onToggleItemExpanded: Function.prototype,
    itemFilters: {},
    visualization: {},
}

const mapStateToProps = (state, ownProps) => ({
    itemFilters: sGetItemFiltersRoot(state),
    visualization: sGetVisualization(
        state,
        pluginManager.extractFavorite(ownProps.item).id
    ),
})

const mapDispatchToProps = dispatch => ({
    onVisualizationLoaded: visualization =>
        dispatch(acAddVisualization(visualization)),
    onSelectActiveType: (id, type, activeType) =>
        dispatch(acSetActiveVisualizationType(id, type, activeType)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Item)
