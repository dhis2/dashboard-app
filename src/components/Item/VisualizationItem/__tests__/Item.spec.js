import React from 'react';
import { shallow } from 'enzyme';
import ChartPlugin from '@dhis2/data-visualizer-plugin';
import { CHART, REPORT_TABLE } from '../../../../modules/itemTypes';
import { Item } from '../Item';
import DefaultPlugin from '../DefaultPlugin';

jest.mock('@dhis2/data-visualizer-plugin', () => () => <div />);
jest.mock('../DefaultPlugin', () => () => <div />);
jest.mock('../ItemFooter', () => () => <div />);
jest.mock('../plugin', () => {
    return {
        getLink: jest.fn(),
        unmount: jest.fn(),
        pluginIsAvailable: () => true,
        getName: () => 'rainbow',
        fetch: () => {},
        getVisualizationConfig: visualization => visualization,
    };
});

describe('VisualizationItem/Item', () => {
    let props;
    let shallowItem;

    const canvas = () => {
        if (!shallowItem) {
            shallowItem = shallow(<Item {...props} />, { context: { d2: {} } });
        }
        return shallowItem;
    };

    beforeEach(() => {
        props = {
            classes: {},
            item: {
                type: REPORT_TABLE,
                id: 'testItem2',
                reportTable: {
                    id: 'pivot1',
                    name: 'Test pivot',
                },
            },
            editMode: false,
            itemFilters: {
                brilliance: [{ id: 100, name: '100' }],
            },
            visualization: {
                name: 'Test pivot',
                description: 'Test pivot mock',
                rows: [],
                columns: [],
                filters: [],
            },
            onToggleItemExpanded: jest.fn(),
            onVisualizationLoaded: jest.fn(),
        };
        shallowItem = undefined;
    });

    it('renders a ChartPlugin when a chart item is passed', () => {
        props.visualization = {
            name: 'Test chart',
            description: 'Test chart mock',
            rows: [],
            columns: [],
            filters: [],
        };
        props.item = {
            type: CHART,
            id: 'testItem1',
            chart: {
                id: 'chart1',
                name: 'Test chart',
            },
        };

        const expectedConfig = {
            ...props.visualization,
            id: undefined,
            filters: [
                {
                    dimension: 'brilliance',
                    items: props.itemFilters.brilliance,
                },
            ],
        };

        const component = canvas();

        component.setState({ configLoaded: true });

        const chartPlugin = component.find(ChartPlugin);

        expect(chartPlugin.exists()).toBeTruthy();
        expect(chartPlugin.prop('config')).toEqual(expectedConfig);
    });

    it('renders a DefaultPlugin when a item different from chart is passed', () => {
        const expectedConfig = {
            ...props.visualization,
            id: undefined,
            filters: [
                {
                    dimension: 'brilliance',
                    items: props.itemFilters.brilliance,
                },
            ],
        };

        expect(canvas()).toMatchSnapshot();

        const component = canvas();

        component.setState({ configLoaded: true });

        const defaultPlugin = canvas().find(DefaultPlugin);

        expect(defaultPlugin.exists()).toBeTruthy();
        expect(defaultPlugin.prop('visualization')).toEqual(expectedConfig);
    });
});
