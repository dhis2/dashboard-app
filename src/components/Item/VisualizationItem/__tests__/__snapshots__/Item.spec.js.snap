// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisualizationItem/Item renders a DefaultPlugin when a item different from chart is passed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Item
    classes={Object {}}
    editMode={false}
    item={
      Object {
        "id": "testItem2",
        "reportTable": Object {
          "id": "pivot1",
          "name": "Test pivot",
        },
        "type": "REPORT_TABLE",
      }
    }
    itemFilters={
      Object {
        "brilliance": Array [
          Object {
            "id": 100,
            "name": "100",
          },
        ],
      }
    }
    onToggleItemExpanded={[MockFunction]}
    onVisualizationLoaded={[MockFunction]}
    visualization={
      Object {
        "columns": Array [],
        "description": "Test pivot mock",
        "filters": Array [],
        "name": "Test pivot",
        "rows": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ItemHeader
          actionButtons={
            <VisualizationItemHeaderButtons
              activeFooter={false}
              activeType="REPORT_TABLE"
              item={
                Object {
                  "id": "testItem2",
                  "reportTable": Object {
                    "id": "pivot1",
                    "name": "Test pivot",
                  },
                  "type": "REPORT_TABLE",
                }
              }
              onSelectVisualization={[Function]}
              onToggleFooter={[Function]}
            />
          }
          editMode={false}
          title={
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <span
                title="rainbow"
              >
                rainbow
              </span>
              <a
                style={
                  Object {
                    "height": 16,
                  }
                }
                title="View in Pivot Tables app"
              >
                <pure(LaunchIcon) />
              </a>
            </div>
          }
        />,
        <div
          className="dashboard-item-content"
        />,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "actionButtons": <VisualizationItemHeaderButtons
            activeFooter={false}
            activeType="REPORT_TABLE"
            item={
              Object {
                "id": "testItem2",
                "reportTable": Object {
                  "id": "pivot1",
                  "name": "Test pivot",
                },
                "type": "REPORT_TABLE",
              }
            }
            onSelectVisualization={[Function]}
            onToggleFooter={[Function]}
          />,
          "editMode": false,
          "title": <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
              }
            }
          >
            <span
              title="rainbow"
            >
              rainbow
            </span>
            <a
              style={
                Object {
                  "height": 16,
                }
              }
              title="View in Pivot Tables app"
            >
              <pure(LaunchIcon) />
            </a>
          </div>,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "dashboard-item-content",
        },
        "ref": [Function],
        "rendered": false,
        "type": "div",
      },
      null,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ItemHeader
            actionButtons={
              <VisualizationItemHeaderButtons
                activeFooter={false}
                activeType="REPORT_TABLE"
                item={
                  Object {
                    "id": "testItem2",
                    "reportTable": Object {
                      "id": "pivot1",
                      "name": "Test pivot",
                    },
                    "type": "REPORT_TABLE",
                  }
                }
                onSelectVisualization={[Function]}
                onToggleFooter={[Function]}
              />
            }
            editMode={false}
            title={
              <div
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                  }
                }
              >
                <span
                  title="rainbow"
                >
                  rainbow
                </span>
                <a
                  style={
                    Object {
                      "height": 16,
                    }
                  }
                  title="View in Pivot Tables app"
                >
                  <pure(LaunchIcon) />
                </a>
              </div>
            }
          />,
          <div
            className="dashboard-item-content"
          />,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "actionButtons": <VisualizationItemHeaderButtons
              activeFooter={false}
              activeType="REPORT_TABLE"
              item={
                Object {
                  "id": "testItem2",
                  "reportTable": Object {
                    "id": "pivot1",
                    "name": "Test pivot",
                  },
                  "type": "REPORT_TABLE",
                }
              }
              onSelectVisualization={[Function]}
              onToggleFooter={[Function]}
            />,
            "editMode": false,
            "title": <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              <span
                title="rainbow"
              >
                rainbow
              </span>
              <a
                style={
                  Object {
                    "height": 16,
                  }
                }
                title="View in Pivot Tables app"
              >
                <pure(LaunchIcon) />
              </a>
            </div>,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "dashboard-item-content",
          },
          "ref": [Function],
          "rendered": false,
          "type": "div",
        },
        null,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "d2": Object {},
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
